src/NATS.Client.KeyValueStore/NatsKVStore.cs

```diff
@@ -315,11 +315,27 @@ public async IAsyncEnumerable<NatsKVEntry<T>> WatchAsync<T>(string key, INatsDes
     {
         await using var watcher = await WatchInternalAsync<T>(key, serializer, opts, cancellationToken);
 
+        if (watcher.InitialConsumer.Info.NumPending == 0 && opts?.EndOfCurrentData != null)
+        {
+            if (await opts.EndOfCurrentData(cancellationToken))
+            {
+                yield break;
+            }
+        }
+
         while (await watcher.Entries.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
         {
             while (watcher.Entries.TryRead(out var entry))
             {
                 yield return entry;
+
+                if (entry.Delta == 0 && opts?.EndOfCurrentData != null)
+                {
+                    if (await opts.EndOfCurrentData(cancellationToken))
+                    {
+                        yield break;
+                    }
+                }
```

Comment:
This could already be done in regular enumeration - checking `entry.Delta` and optionally running a funciton

Reply:
@darkwatchuk, I'm going to remove this bit if that's ok with you? We can always check this in the loop as @caleblloyd suggested, and also callback can be just for one thing; there was no data.